### ISSUE TITLE
Struggling to get the labels for visual, audio, and text annotated at each sentence as described in the paper

### ISSUE THREAD

**Issue by e-remington-lee** [A2Zadeh/CMU-MultimodalSDK#242]
Hello!

I am running on a Windows OS and I am having some trouble with the SDK so I opted with using the raw data and trying to manually set up the data to run experiments. What I am trying to do right now is to get the labels on a pre-sentence basis for the visual,  audio, and text data. The goal would be if I can get the visual, audio, and text emotion and sentiment labels for each sentence in each video I could go back and match the segments of the videos with those labels and begin extracting my own features from the data. I would assume I would need to extract MFCC's from the audio data, the facial coordinates from the videos at each frame, then match the sentences to those features to provide the labels.

I see here, http://immortal.multicomp.cs.cmu.edu/raw_datasets/processed_data/cmu-mosei/seq_length_50/, that there is pickled data that I can extract which has a dictionary with"train, validation, and test" keys, each with "vision, audio, labels, and id" keys. Theer are 22777 different values too, so I assume these are all of the sentences in the dataset. The labels make sense because they are in a range between [-3,3] which denotes the 7 different emotions, but I don't understand what the vision, audio, text, and ID columns represent. The vision, audio, and text data are confusing because I am not sure what the numbers represent (pixels? mfcc data? I do not know) and the ID column is confusing because I do not understand how to link each sentence back to which video it is related to and at what time it is related to as well.

I have been scouring the resolved issues to see if there are any answers to my questions, but I cannot seem to find anything. I found https://github.com/A2Zadeh/CMU-MultimodalSDK/issues/241 which talks about a similar issue but did not really answer any of the questions I have.

If there are other issues that have been resolved that are similar to this one, feel free to link me to those! I have been struggling for a few weeks to get this to work because I cannot find a lot of information about labeling all the data and aligning them with everything else.

**Comment by A2Zadeh** [comment]
Hi @e-remington-lee,

Happy to help. Let's do a quick check before we dive into the solution. The way you have opted to use the data will make your work much harder overall. So I wanna see if we can find a way forward for you to use the data easily.

1. Are you familiar with word alignment done between modalities?
2. Have you looked at the SDK examples? It might help. 
3. Have you been able to run a baseline on the dataset? I assume you are interested in MulT by Hubert Tsai, since that is the data you referred to.

**Comment by e-remington-lee** [comment]
Hello @A2Zadeh 

Thanks for the response. Here are the answers to your questions:

1. I am not familiar with word alignment between modalities. The AI/ML projects I have done in the past have only relied on 1 modality (image classification, NLP sentiment analysis) so I was hoping to learn how to successfully work with a multi-modal dataset. The reason I wanted to opt for processes the features myself is that I want to learn how you got to the data you did so I am able to replicate this on a potential new project in the future. Understanding the end-to-end process helps me learn a lot better than just understanding a piece of it. I understand it might be exponentially more difficult, but I believe what I will learn along the way is what is important.

2. I have looked at some of the examples. I keep trying to run the example here "https://github.com/A2Zadeh/CMU-MultimodalSDK/blob/master/examples/mmdatasdk_examples/full_examples/process_mosei.py" but it keeps stalling out at "Pre-alignment based on <glove_vectors> computational sequence started..." It is possible I need to wait longer, I will let it run for a few hours after submitting this comment. I am also unsure exactly what it is doing and what I will have accomplished once the program succeeds. I see it says "processed data" but what does it mean to process data? Did we extract the features from the videos or are we organizing the data in a way we can run AI/ML models.

3. I have not been able to run a baseline on the dataset because I am unable to understand what data I am looking at. In my past experience working with datasets, there is a readme file that describes what each column in the data is and the data is typically in some text file format so I can easily view it. My issue at the moment is I do not understand what the numbers in the data represent (pixels, or some other feature), I do not understand how to correlate the processed data back to the raw unprocessed data, and I am having trouble viewing the data in a readable way (similar to a text file). As a personal preference, I like to be able to look at the labeled data and cross-check it with what I see in the video example myself, that way I can be very confident in exactly what the data is showing me and I can move forward.

I hope my answers were satisfactory and helped you understand my perspective and where I am at.

**Comment by A2Zadeh** [comment]
Hi @e-remington-lee,

1. It's great that you are planning to get involved in multimodal research. It is a very friendly and diverse field. I hope you find our datasets helpful to your research. Please have a look at this paper, where we first introduced word alignment: https://dl.acm.org/doi/abs/10.1145/3136755.3136801  
Subsequent papers such as MFN make things a bit more clear. 

2. Yes, alignment is a heavy process. I have tried to make it as fast as possible, but still it is a significant process. The pre-alignment makes everything ready for a quick alignment to follow. You can extract features but since this is among the first interactions you have with multimodal sequential data, I don't advocate that. It's easier to start from preprocessed stuff. 

3. Information regarding feature dimensions are available in the "dimension names" of the metadata for computational sequences. We also have that information in the papers. We don't usually deal with pixels because we easily overfit - say, blue background means positive sentiment 100%. Paul has a good tutorial on MFN: https://github.com/pliang279/MFN and Zhun has a very good one on https://github.com/Justin1904/TensorFusionNetworks

Please let me know if you had other questions.

**Comment by e-remington-lee** [comment]
Hey @A2Zadeh 

Thanks for taking the time to respond, your response was extremely helpful! I will reach out again if I have other questions.
